galley/cmd/gals/cmd/validator.go

```diff
@@ -0,0 +1,149 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"istio.io/istio/galley/cmd/shared"
+	"istio.io/istio/galley/pkg/crd/validation"
+	"istio.io/istio/pilot/pkg/bootstrap"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/util"
+)
+
+// waitSignal awaits for SIGINT or SIGTERM and closes the channel
+func waitSignal(stop chan struct{}) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	<-sigs
+	close(stop)
+	_ = log.Sync()
+}
+
+func validatorCmd(printf, fatalf shared.FormatFn) *cobra.Command {
+	var (
+		webhookConfigName   string
+		pilotWebhookName    string
+		mixerWebhookName    string
+		port                uint
+		domainSuffix        string
+		certFile            string
+		keyFile             string
+		caFile              string
+		healthCheckInterval time.Duration
+		healthCheckFile     string
+	)
+
+	patchCert := func(stop chan struct{}) error {
+		caCertPem, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return err
+		}
+
+		timeout := time.After(time.Minute)
+		cl := common.client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations()
+		for {
+			webhooks := []string{pilotWebhookName, mixerWebhookName}
+			if err = util.PatchValidatingWebhookConfig(cl, webhookConfigName, webhooks, caCertPem); err == nil {
+				return nil
+			}
+
+			log.Errorf("Patching caBundle in ValidatingWebhookConfigurations %v failed: %v -- retrying",
+				webhookConfigName, err)
+			select {
+			case <-stop:
+				// canceled by caller
+				return nil
+			case <-time.After(time.Second):
+				// retry
+			case <-timeout:
+				return fmt.Errorf("timed out trying to patch ValidatingWebhookConfiguration %q {%q and %q})",
+					webhookConfigName, pilotWebhookName, mixerWebhookName)
+			}
+		}
+	}
+
+	validatorCmd := &cobra.Command{
+		Use:   "validator",
+		Short: "Runs an https server for Istio configuration validation.",
+		Long: "Runs an https server for Istio configuration validation. " +
+			"Uses k8s validating webhooks to validate Pilot and Mixer configuration.",
+		Run: func(cmd *cobra.Command, args []string) {
+			mixerValidator, err := validation.CreateMixerValidator(flags.kubeConfig)
+			if err != nil {
+				fatalf("cannot create mixer backend validator for %q: %v", flags.kubeConfig, err)
+			}
+
+			params := validation.WebhookParameters{
+				MixerValidator:      mixerValidator,
+				PilotDescriptor:     bootstrap.ConfigDescriptor,
+				DomainSuffix:        domainSuffix,
+				Port:                port,
+				CertFile:            certFile,
+				KeyFile:             keyFile,
+				HealthCheckInterval: healthCheckInterval,
+				HealthCheckFile:     healthCheckFile,
+			}
+			wh, err := validation.NewWebhook(params)
+			if err != nil {
+				fatalf("cannot create validation webhook service: %v", err)
+			}
+
+			// Create the stop channel for all of the servers.
+			stop := make(chan struct{})
+
+			go func() {
+				if err := patchCert(stop); err != nil {
```

Comment:
What happens when patchCert fails? It looks like the  webhook will still try to serve?

Reply:
Validation will be broken. The webhook will continue to serve but api-server won't call it. The sidecar injector has a similar problem. I'm hoping the environments WG can address this problem as part of k8s install/upgrade process. cc @yusuoh

Reply:
sidecar injector service will failed if it cannot patch the cert after several retries.

https://github.com/istio/istio/blob/master/pilot/cmd/sidecar-injector/main.go#L86